File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/View/OpenGLTest.java
package com.spvessel.spacevil.View;

import java.util.*;
import java.awt.Color;
import java.awt.image.BufferedImage;
import com.spvessel.spacevil.*;
import com.spvessel.spacevil.Common.*;
import com.spvessel.spacevil.Core.*;
import com.spvessel.spacevil.Flags.*;
import com.spvessel.spacevil.Decorations.*;

public class OpenGLTest extends ActiveWindow {
   @Override
   public void initWindow() {
      setParameters(this.getClass().getName(), this.getClass().getName(), 800, 800, false);
      isCentered = true;
      isMaximized = true;

      // oneCubeExample();
      multipleCubes();
   }

   private void oneCubeExample() {
      TitleBar titleBar = new TitleBar(this.getClass().getName());

      OpenGLLayer ogl = new OpenGLLayer();
      ogl.setMargin(0, titleBar.getHeight(), 0, 0);

      HorizontalStack toolbar = getToolbarLayout();

      ImagedButton btnRotateLeft = getImagedButton(EmbeddedImage.ARROW_UP, -90);
      ImagedButton btnRotateRight = getImagedButton(EmbeddedImage.ARROW_UP, 90);

      HorizontalSlider zoom = getSlider();

      ImagedButton btnRestoreView = getImagedButton(EmbeddedImage.REFRESH, 0);

      // adding
      addItems(titleBar, ogl);
      ogl.addItems(toolbar);
      toolbar.addItems(btnRotateLeft, btnRotateRight, zoom, btnRestoreView);

      // assign events
      btnRestoreView.eventMousePress.add((sender, args) -> {
         ogl.restoreView();
      });

      btnRotateLeft.eventMousePress.add((sender, args) -> {
         ogl.rotate(KeyCode.LEFT);
      });

      btnRotateRight.eventMousePress.add((sender, args) -> {
         ogl.rotate(KeyCode.RIGHT);
      });

      zoom.eventValueChanged.add((sender) -> {
         ogl.setZoom(zoom.getCurrentValue());
      });

      // set focus
      ogl.setFocus();
      zoom.setCurrentValue(3);
   }

   private void multipleCubes() {
      TitleBar titleBar = new TitleBar(this.getClass().getName());

      FreeArea area = new FreeArea();
      area.setMargin(0, titleBar.getHeight(), 0, 0);

      addItems(titleBar, area);

      List<InterfaceBaseItem> content = new ArrayList<>();

      for (int row = 0; row < 3; row++) {
         for (int column = 0; column < 3; column++) {
            ResizableItem frame = new ResizableItem();
            frame.setBorder(new Border(Color.gray, new CornerRadius(), 2));
            frame.setPadding(5, 5, 5, 5);
            frame.setBackground(100, 100, 100);
            frame.setSize(200, 200);
            frame.setPosition(90 + row * 210, 60 + column * 210);
            area.addItem(frame);
            content.add(frame);

            frame.eventMousePress.add((sender, args) -> {
               content.remove(frame);
               content.add(frame);
               area.setContent(content);
            });

            OpenGLLayer ogl = new OpenGLLayer();
            ogl.setMargin(0, 30, 0, 0);
            frame.addItem(ogl);
         }
      }
   }

   public static ImagedButton getImagedButton(EmbeddedImage image, float imageRotationAngle) {
      ImagedButton btn = new ImagedButton(DefaultsService.getDefaultImage(image, EmbeddedImageSize.SIZE_32X32),
            imageRotationAngle);
      return btn;
   }

   public static HorizontalSlider getSlider() {
      HorizontalSlider slider = new HorizontalSlider();
      slider.setStep(0.2f);
      slider.setMinValue(2f);
      slider.setMaxValue(10f);
      slider.setSizePolicy(SizePolicy.FIXED, SizePolicy.FIXED);
      slider.setSize(150, 30);
      slider.setAlignment(ItemAlignment.HCENTER, ItemAlignment.BOTTOM);
      slider.setMargin(40, 0, 40, 0);
      slider.handler.setShadow(5, 0, 2, Color.BLACK);
      return slider;
   }

   public static HorizontalStack getToolbarLayout() {
      HorizontalStack layout = new HorizontalStack();
      layout.setContentAlignment(ItemAlignment.HCENTER);
      layout.setAlignment(ItemAlignment.BOTTOM, ItemAlignment.LEFT);
      layout.setSizePolicy(SizePolicy.EXPAND, SizePolicy.FIXED);
      layout.setHeight(30);
      layout.setSpacing(5, 0);
      layout.setMargin(20, 0, 20, 20);
      return layout;
   }

   public static class ImagedButton extends BlankItem {
      ImageItem _image = null;

      public ImagedButton(BufferedImage image, float imageRotationAngle) {
         addItemState(ItemStateType.HOVERED, new ItemState(new Color(255, 255, 255, 40)));
         addItemState(ItemStateType.PRESSED, new ItemState(new Color(0, 0, 0, 40)));
         setBackground(0xFF, 0xB5, 0x6F);
         setSizePolicy(SizePolicy.FIXED, SizePolicy.FIXED);
         setSize(30, 30);
         setAlignment(ItemAlignment.RIGHT, ItemAlignment.BOTTOM);
         setPadding(5, 5, 5, 5);
         setShadow(5, 0, 2, Color.BLACK);
         isFocusable = false;

         _image = new ImageItem(image, false);
         _image.setRotationAngle(imageRotationAngle);
         _image.setColorOverlay(Color.WHITE);
      }

      @Override
      public void initElements() {
         super.initElements();
         addItem(_image);
      }
   }
}

File: CSharp/SpaceVIL/SpaceVIL/View/OpenGLTest.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using SpaceVIL;
using SpaceVIL.Common;
using SpaceVIL.Core;
using SpaceVIL.Decorations;

namespace View
{
    public class OpenGLTest : ActiveWindow
    {
        public override void InitWindow()
        {
            SetParameters(this.GetType().Name, this.GetType().Name, 800, 800, false);
            IsCentered = true;
            IsMaximized = true;

            // OneCubeExample();
            // MultipleCubes();
        }

        // private void OneCubeExample()
        // {
        //     TitleBar titleBar = new TitleBar(this.GetType().Name);

        //     OpenGLLayer ogl = new OpenGLLayer();
        //     ogl.SetMargin(0, titleBar.GetHeight(), 0, 0);

        //     HorizontalStack toolbar = GetToolbarLayout();

        //     ImagedButton btnRotateLeft = GetImagedButton(EmbeddedImage.ArrowUp, -90);
        //     ImagedButton btnRotateRight = GetImagedButton(EmbeddedImage.ArrowUp, 90);

        //     HorizontalSlider zoom = GetSlider();

        //     ImagedButton btnRestoreView = GetImagedButton(EmbeddedImage.Refresh, 0);

        //     // adding
        //     AddItems(titleBar, ogl);

        //     AddItems(toolbar);
        //     toolbar.AddItems(btnRotateLeft, btnRotateRight, zoom, btnRestoreView);

        //     // assign events
        //     btnRestoreView.EventMousePress += (sender, args) =>
        //     {
        //         ogl.RestoreView();
        //     };

        //     btnRotateLeft.EventMousePress += (sender, args) =>
        //     {
        //         ogl.Rotate(KeyCode.Left);
        //     };

        //     btnRotateRight.EventMousePress += (sender, args) =>
        //     {
        //         ogl.Rotate(KeyCode.Right);
        //     };

        //     zoom.EventValueChanged += (sender) =>
        //     {
        //         ogl.SetZoom(zoom.GetCurrentValue());
        //     };

        //     // Set focus
        //     ogl.SetFocus();
        //     zoom.SetCurrentValue(3);
        // }

        // private void MultipleCubes()
        // {
        //     TitleBar titleBar = new TitleBar(this.GetType().Name);

        //     FreeArea area = new FreeArea();
        //     area.SetMargin(0, titleBar.GetHeight(), 0, 0);

        //     AddItems(titleBar, area);

        //     List<IBaseItem> content = new List<IBaseItem>();

        //     for (int row = 0; row < 3; row++)
        //     {
        //         for (int column = 0; column < 3; column++)
        //         {
        //             ResizableItem frame = new ResizableItem();
        //             frame.SetBorder(new Border(Color.Gray, new CornerRadius(), 2));
        //             frame.SetPadding(5, 5, 5, 5);
        //             frame.SetBackground(100, 100, 100);
        //             frame.SetSize(200, 200);
        //             frame.SetPosition(90 + row * 210, 60 + column * 210);
        //             area.AddItem(frame);
        //             content.Add(frame);

        //             frame.EventMousePress += (sender, args) =>
        //             {
        //                 content.Remove(frame);
        //                 content.Add(frame);
        //                 area.SetContent(content);
        //             };

        //             OpenGLLayer ogl = new OpenGLLayer();
        //             ogl.SetMargin(0, 30, 0, 0);
        //             frame.AddItem(ogl);
        //         }
        //     }
        // }

        public static ImagedButton GetImagedButton(EmbeddedImage image, float imageRotationAngle)
        {
            ImagedButton btn = new ImagedButton(DefaultsService.GetDefaultImage(image, EmbeddedImageSize.Size32x32),
                    imageRotationAngle);
            return btn;
        }

        public static HorizontalSlider GetSlider()
        {
            HorizontalSlider slider = new HorizontalSlider();
            slider.SetStep(0.2f);
            slider.SetMinValue(2f);
            slider.SetMaxValue(10f);
            slider.SetSizePolicy(SizePolicy.Fixed, SizePolicy.Fixed);
            slider.SetSize(150, 30);
            slider.SetAlignment(ItemAlignment.HCenter, ItemAlignment.Bottom);
            slider.SetMargin(40, 0, 40, 0);
            slider.Handler.SetShadow(5, 0, 2, Color.Black);
            return slider;
        }

        public static HorizontalStack GetToolbarLayout()
        {
            HorizontalStack layout = new HorizontalStack();
            layout.SetContentAlignment(ItemAlignment.HCenter);
            layout.SetAlignment(ItemAlignment.Bottom, ItemAlignment.Left);
            layout.SetSizePolicy(SizePolicy.Expand, SizePolicy.Fixed);
            layout.SetHeight(30);
            layout.SetSpacing(5, 0);
            layout.SetMargin(20, 0, 20, 20);
            return layout;
        }
    }

    public class ImagedButton : ButtonCore
    {
        ImageItem _image = null;

        public ImagedButton(Bitmap image, float imageRotationAngle)
        {
            SetBackground(0xFF, 0xB5, 0x6F);
            SetSizePolicy(SizePolicy.Fixed, SizePolicy.Fixed);
            SetSize(30, 30);
            SetAlignment(ItemAlignment.Right, ItemAlignment.Bottom);
            SetPadding(5, 5, 5, 5);
            SetShadow(5, 0, 2, Color.Black);
            IsFocusable = false;

            _image = new ImageItem(image, false);
            _image.SetRotationAngle(imageRotationAngle);
            _image.SetColorOverlay(Color.White);
        }

        public override void InitElements()
        {
            base.InitElements();
            AddItem(_image);
        }
    }
}